### DISCUSSION TITLE
Library Of Plugin Libraries Is Needed - CENTRALISED

### DISCUSSION THREAD

**Discussion by TiddlyTweeter** [TiddlyWiki/TiddlyWiki5#5524] [Ideas]
### We have NO Central Library Of Plugin Libraries. AND We Need It.  ??

I think a major issue in promoting TW is that we have _NO Library Of Plugin Libraries_ . 

Plugins give much utility to TW. But the system, as of now, is down to individuals to find and install linkage to Plugin Libraries.

That seems WRONG. 

FYI, my plugin Library of Libraries is these Tiddlers ...

![Screenshot 2021-03-05 122443](https://user-images.githubusercontent.com/28830541/110110223-cce0d400-7dae-11eb-8f2c-feeca983ddf5.jpg)

 I DO understand that TW is UN-centralised. But on this issue I do think that a method to REGISTER A LIBRARY CENTRALLY would help us a lot promote TW.

A case-in-point is SOUNDCITE that has a Plugin Library. 
If I have never mentioned it would YOU know it existed?

**Comment by twMat** [comment]
There is a [general agreement and discussion](https://github.com/Jermolene/TiddlyWiki5/discussions/5399) on the need for this but there is not (yet?) enough incentive/knowledge/time to actually create it. That discussion is not particularly about what you request here but if the referenced issue is solved then your question would be simple to solve. 

However, @Marxsal and I are building a Google Sheets plugin and one of *my* incentives for this is to solve the problem. For example, there could be a google form that people can use to post links to plugins which is then automatically collected in a central list. The/A "google sheets plugin" could be designed to import a tiddler that lists all table entries along with tags or other meta data. 

It is unfortunate to rely on Google for this basic need of ours but this is what it's coming to after me needing it for 15 years, but not being able to solve it.

**Reply by Marxsal** [reply]
> For example, there could be a google form that people can use to post links to plugins which is then automatically collected in a central list.

Isn't that what tiddly toolmap does?

**Reply by twMat** [reply]
>Isn't that what tiddly toolmap does?

Yes! But there are at least two significant problems with Tiddly Toolmap, like all similar alternatives previously (even if Davids is definitely the most successful so far):

* It suffers from the lonewolf syndrome; It relies on a single guy to host and curate it. If David abandons it (or even just takes a break) it starts to decay.
* It is not a TW plugin or integrated with TW in any way.

That last point is critical: There is only way to *reliably, consistently, and without effort* draw attention to someting in TiddlyVerse and that is to have it prominently presented on tiddlywiki.com. (A mere tiddler in the Community resources is in no way enough.) The only "external" thing ever promoted in this way is Felix' Tiddly*Map* (and TWC). Without this main spot it is a Sisyphus struggle to keep make people use something. **But** a plugin at least partly get around this because if they just install it *once* in their TW, it is there and keeps reminding them when they look at their plugins in their own TW.

**Reply by Marxsal** [reply]
I'm trying to imagine how this would work. Are you saying that there would be one central spreadsheet that **anyone** could contribute to? In which case, how do you prevent ne'r-do-wells from spamming or destroying it? 

Or are you saying that there will be various spreadsheets set up by various (mostly) individual collectors, and then one central TiddlyGoo that knows those spreadsheet IDs so that people can download their own collection library? Kind of like twederation (for which the logic was never worked out) but based on Google spreadsheets?

Hmm. Maybe that would work. I guess that central TG would also need a list of known hashes so it could verify any plugins pulled in.

**Reply by twMat** [reply]
>Or are you saying that there will be various spreadsheets set up by various (mostly) individual collectors, and then one central TiddlyGoo that knows those spreadsheet IDs so that people can download their own collection library? Kind of like twederation (for which the logic was never worked out) but based on Google spreadsheets?

@Marxsal 

OK, I have many vague ideas about this because the sheets are more complex than what one might at first think and for this reason I think this can grow in ways we don't expect. For example: Spreadsheets can contain *different kinds of data* such as*actual* tiddlers and/or *links* to tiddlers and/or links to *other* spreadsheets. Further, consider that spreadsheets are not limited to pure tiddly stuff. E.g as an elementary school science teacher, I'm just fiddling on a system so that I can create dynamic assignment papers by fetching data from GDrive into TW. ("I want: 10 questions, physics, difficulty B, grade 5") and if I find other teachers then I'm sure we could share other stuff via the TiddlyGoo infrastructure and public sheets. (I've had ideas along these lines [for many years](http://twaddle.tiddlyspot.com/#Different%20worlds) many years.) Further, spreadsheets can be scripted to fetch data from other spreadsheets (maybe [from wikis](https://github.com/Jermolene/TiddlyWiki5/discussions/5525)), even automatically, so there can be a big spread of content among sheets. My point is that we can talk about "one central" etc but maybe nobody is interested in "the central" that we had in mind.

BUT we do have some control of this as the creators of *the link* between TW and Sheets: The very SheetsIN/OUT plugin (be it one or two) can, for example, provide a UI to make it easy for end users to share the IDs of sheets (really SpreadsheetTiddlers) that they want to share to a central list. 

To prevent from bad apples I think a rating system is needed. And, of course, the rating has to be tied to an identity - which in this case has to mean the spreadsheet IDs. We could have a system where people rate creations (really their IDs) and the results are presented on the central list. (Rating something is possible via google forms).

OK, this is getting too intricate. I think the best we can do ATM is to create the link that enables it all. Later, changes will definitely come  :-)

**Reply by TiddlyTweeter** [reply]
> @twMat There is a [general agreement and discussion](https://github.com/Jermolene/TiddlyWiki5/discussions/5399) on the need for this but there is not (yet?) enough incentive/knowledge/time to actually create it. That discussion is not particularly about what you request here but if the referenced issue is solved then your question would be simple to solve.

Right. In my OP I was concentrating more on a more minimal case: **Plugin Libraries** only. But similar issues apply. I simply think that extant Plugin Libraries are a clear use case where "centralisation" more obviously makes sense. And at a SCALE that is likely realisable.

**Reply by TiddlyTweeter** [reply]
> @Marxsal  I'm trying to imagine how this would work. Are you saying that there would be one central spreadsheet that **anyone** could contribute to? In which case, how do you prevent ne'r-do-wells from spamming or destroying it?

Right. Spamming inside TW work is currently like zero. But I guarantee it will become an issue IF we got "gooder" at our good. 

Your observation is THAT whole problem of "centralisation" in a nut. 

The OP was mainly focused on a delimited case. To give end-users a bit more utility through a small collection of **Plugin Libraries** pieces. That itself raises enough issues already. _WHAT does a specific plugin within one DO?_  :-) HOW would we know?

TT

**Comment by pmario** [comment]
If there is some "centralized" - "load from library" backend I personally expect it to be **secure**. At the moment we don't have any security mechanisms in place. 

The only security comes from trust. We trust the "core plugins". Some creators use their own libraries. ... Users use them, because they trust them. ... BUT if there is a centralized repository of libraries, a single user isn't able anymore to be "up to date" which library and plugin is save to be included. 

## Proposal

The minimal security measure that I can think of is a "hash", which will be automatically created by the central library and it will be published there. 

The core "import mechanism" should check every plugin against this hash and tell the user if there are any differences. ... 

BUT

This mechanism alone isn't enough, but it would be a start!

----

The "hash" can only tell the user, that the plugin version, which is part of the wiki, is the same as the plugin that was published in the central repo.

**Reply by TiddlyTweeter** [reply]
> @pmario  If there is some "centralized" - "load from library" backend I personally expect it to be **secure**. At the moment we don't have any security mechanisms in place.

I have NO understanding of how you do that VALIDATION. But fully accept it is needed.

FWIW, my OP is limited to only thoughts on somehow AGGREGATING a **Library Of Plugin Libraries**. _Why? Because they exist._ Are very useful. Tend to be by trusted developers. Are currently finite in number (meaning one can approach "issues" in a controlled way). Yet they DO, already, raise the issue of meta-indexing: _How would a user know which plugins are relevant to them in a GROUP of libraries?_ There is no meta-index of application/plugin remit as far as I know?

Thoughts, TT

**Comment by Jermolene** [comment]
I think the OP is asking for a central registry of plugin libraries, which is a fairly simple thing that doesn't by itself raise the kind of issues we've discussed over on #5399. The obvious way of achieving it would be to add a new tab of "libraries" to the "add plugins" modal, so that users could choose to install the plugin libraries they need. The user experience would be pretty convoluted for end users, so I'm not sure whether it's worth going that far.

But surely we could have a register of plugin libraries on tiddlywiki.com, so that people can drag and drop from there. That's pretty easy: we'd add a new tiddler "Community Plugin Libraries" that lists individual `$:/tags/PluginLibrary` entries, and make it prominently linked in the TOC under "Community".

**Reply by pmario** [reply]
>But surely we could have a register of plugin libraries on tiddlywiki.com, so that people can drag and drop from there. That's pretty easy: we'd add a new tiddler "Community Plugin Libraries" that lists individual $:/tags/PluginLibrary entries, and make it prominently linked in the TOC under "Community".

+1

**Reply by TiddlyTweeter** [reply]
> @Jermolene ... I think the OP is asking for a central registry of plugin libraries, which is a fairly simple thing that doesn't by itself raise the kind of issues we've discussed over on #5399. 

Right. **Spot-on.** 

I simply thought that **Plugin Libraries** are a limited set with maximal utility that direct ability to install would be of high benefit at low overhead. It seems potentially easy now.

We can leave till later the fact that just installing a library doesn't add to any index of _"what is in it?"_ That can wait till its many more libraries. But it will need to come, I think?

My thoughts,
Ciao, ciao

**Reply by Jermolene** [reply]
> We can leave till later the fact that just installing a library doesn't add to any index of _"what is in it?"_ That can wait till its many more libraries. But it will need to come, I think?

Users would indeed need to explicitly open each plugin library to inspect the contents. One could imagine making a button that retrieved the listings from all the available plugin libraries. I think it would probably be more useful to work on the centralised community plugin library but there's nothing to stop others hacking on that kind of thing as it is all done in wikitext.
